realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -75,6 +84,82 @@ static RealmCacheType valueOf(Class<? extends BaseRealm> clazz) {
+            this.configuration = configuration;
+            this.realmClass = realmClass;
+            this.callback = callback;
+            this.notifier = notifier;
+        }
+
+        @Override
+        public void run() {
+            T instance = null;
+            final AtomicBoolean taskCanceled = new AtomicBoolean(false);
+            try {
+                instance = createRealmOrGetFromCache(configuration, realmClass);
+                boolean results = notifier.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        // If the RealmAsyncTask.cancel() is called before, we just return without creating the Realm
+                        // instance on the caller thread.
+                        // Thread.isInterrupted() cannot be used for checking here since CountDownLatch.await() will
+                        // will clear interrupted status.
+                        if (taskCanceled.get()) {
+                            return;
+                        }
+                        T instanceToReturn = null;
+                        try {
+                            instanceToReturn = createRealmOrGetFromCache(configuration, realmClass);
+                        }catch (RuntimeException e) {
+                            callback.onError(e);
+                        } finally {
+                            fgRealmCreatedLatch.countDown();
+                        }
+                        if (instanceToReturn != null) {
+                            callback.onSuccess(instanceToReturn);
+                        }
+                    }
+                });
+                if (!results) {
+                    fgRealmCreatedLatch.countDown();
+                }
+                // There is a small chance that the posted runnable cannot be executed because of the thread terminated
+                // before the runnable gets fetched from the event queue.
+                if (!fgRealmCreatedLatch.await(5, TimeUnit.SECONDS)) {
+                    RealmLog.warn("Timeout for creating Realm instance in foreground thread in `CreateRealmRunnable` ");
+                }
+            } catch (InterruptedException e) {
+                taskCanceled.set(true);
+                RealmLog.warn(e, "`CreateRealmRunnable` has been interrupted.");
+            } catch (final RuntimeException e) {
+                RealmLog.error(e, "`CreateRealmRunnable` failed.");
+                notifier.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        callback.onError(e);
+                    }
+                });
+            } finally {
+                if (instance != null) {
+                    instance.close();
+                }
+            }
+        }
+    }
+
+    private static final String ASYNC_NOT_ALLOWED_MSG =
+            "The Realm instance cannot be got asynchronously on current thread.";
```

Comment:
Something is strange about `cannot be got`, also the error message doesn't specify *why*.